Add AES-XTS assembly implementation and proofs

pennyannn opened the pull request:
*Description of changes:*

This PR contains work done jointly by @nebeid  and @pennyannn, it:
1. Adds two assembly implementations: `aes_hw_xts_encrypt` and `aes_hw_xts_decrypt`.
2. Adds 20 KAT for each function, these tests are gotten from AWS-LC.
3. Adds the specification and correctness proofs for both functions. 
4. Adds two simple proofs of `aes_encrypt` and `aes_decrypt` in the tutorial. These proofs are the by-product of the authors' initial experiments exploring proving theorems using HOL-Light and s2n-bignum. We think these would be helpful for future contributors.
5. Adds collected signatures for both functions.
6. Adds benchmarking.
7. Adds missing sematest for `strb` and `ldrb`

Some benchmark results on my local laptop (Apple M4):

aes_xts_encrypt (16 bytes)      :    12.5 ns each (var  5.5%, corr -0.04) =   79964570 ops/sec
aes_xts_encrypt (32 bytes)      :    13.9 ns each (var  5.4%, corr -0.01) =   71901064 ops/sec
aes_xts_encrypt (64 bytes)      :    15.2 ns each (var  6.5%, corr -0.06) =   65937633 ops/sec
aes_xts_encrypt (128 bytes)     :    19.3 ns each (var  3.6%, corr  0.11) =   51873015 ops/sec
aes_xts_encrypt (256 bytes)     :    32.2 ns each (var  3.9%, corr -0.20) =   31095420 ops/sec
aes_xts_encrypt (512 bytes)     :    48.5 ns each (var  3.6%, corr -0.10) =   20632960 ops/sec

aes_xts_decrypt (16 bytes)      :    12.6 ns each (var  4.4%, corr  0.14) =   79652958 ops/sec
aes_xts_decrypt (32 bytes)      :    14.4 ns each (var  4.1%, corr  0.28) =   69618491 ops/sec
aes_xts_decrypt (64 bytes)      :    15.8 ns each (var  4.5%, corr -0.06) =   63355297 ops/sec
aes_xts_decrypt (128 bytes)     :    19.8 ns each (var  5.3%, corr -0.12) =   50495634 ops/sec
aes_xts_decrypt (256 bytes)     :    32.7 ns each (var  3.1%, corr  0.14) =   30601196 ops/sec
aes_xts_decrypt (512 bytes)     :    49.3 ns each (var  5.4%, corr  0.04) =   20274561 ops/sec

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

pennyannn committed:
aes-xts experiments

pennyannn committed:
More experiments

pennyannn committed:
More experiments

nebeid committed:
Add Decrypt function- its instructions are all understood by the decoder.

"Remove `.size` as not passing on MacOS with this build.

nebeid committed:
The decrypt version is now from linux-aarch64.

It's from AWS-LC's
generated-src/linux-aarch64/crypto/fipsmodule/aesv8-armx.S
It was previously taken from win-aarch64.

nebeid committed:
enable stopping at first error.

nebeid committed:
Save decoded output to a text file.

nebeid committed:
Update decoded output after adding post-index register offset to the decoder.

nebeid committed:
New decoded output after shuffling around some instructions.

In some places, only comments are different after the decoder was updated.

pennyannn committed:
Start proof for AES256 encrypt

nebeid committed:
Continuing the AES specification.

nebeid committed:
The spec test vector succeeds after converting the inputs to little endian representation.

pennyannn committed:
Add aes-xts decrypt programs

nebeid committed:
Post conditions and MAYCHANGE for an AES block encryption proof.

pennyannn committed:
Add aes program and spec

pennyannn committed:
Symbolic simulation and fixed a bug in code

pennyannn committed:
Finish the proof for aes-xts-decrypt

pennyannn committed:
Set up for aes-xts one block proof

nebeid committed:
Proof of AES encryption of one block.

pennyannn committed:
Fix a bug

pennyannn committed:
Define spec for one block xts and define conversion

pennyannn committed:
Add proof for one block xts

nebeid committed:
One block of AES-256-XTS encrypt: Spec and a test vector (using conversions).

nebeid committed:
Adding new assembly for AES-XTS:

- after merging pennyann's changes to the aws-lc branch
https://github.com/nebeid/aws-lc/commits/aes-xts-slothy/
in commit 8a48d97 which have an updated cleaner assembly

- and cherry picking pennyann's commit Add LDP/STP pre-index and post-index
https://github.com/nebeid/s2n-bignum/commit/182bc754c558805a1ddcb959bb3896d037091dde

The assembly was tested on M3 and was disassembled there and added in this commit.

nebeid committed:
Proof of one-block AES-XTS encrypt.

pennyannn committed:
Update one block proof to use more clean version

pennyannn committed:
WIP: define XTS full specification

pennyannn committed:
Experiment with conversions

nebeid committed:
AES-XTS encrypt (any input length) spec. To be tested

pennyannn committed:
Fix type issue in cipher stealing

pennyannn committed:
Write conversions and add tests

pennyannn committed:
The ultimate clean up

pennyannn committed:
The ultimate ultimate clean up to fix perror

pennyannn committed:
Update clean version of xts decrypt

pennyannn committed:
Update the specification to match the standard for tweak calculation

pennyannn committed:
Add a structure for the proof

nebeid committed:
Make encrypt_spec match decrypt_spec.

nebeid committed:
Working through the conversions for AES-XTS encrypt test vector.

nebeid committed:
Fixed a bug in cipher_stealing_encrypt in the computation of the last tweak.

It was using the input iv value, while it should be using the tweak
that was calculated in the step before using calculate_tweak.

Used the test vectors used in aes_xts_decrypt_spec.ml after modifying
the key schedule of the data key to be as should be used in encryption.
The one used in decryption is different so as to accommodate EQINVCIPHER;
a more efficient decryption algorithm defined in FIPS 197 and used in
AWS-LC.

nebeid committed:
Completed AES-XTS Encrypt specs and their tests.

pennyannn committed:
Eliminating some CHEAT_TACs

pennyannn committed:
Eliminate more CHEAT_TACs and prove a bunch of lemmas

nebeid committed:
Use the asm code integrated in AWS-LC and redo the 1-block proof based on the new specs 1-block function.

pennyannn committed:
Fixing MAYCHANGE and preserved registers

nebeid committed:
WIP: full proof: - loop invariant, one subgoal proven. - proof of code before the loop started.

pennyannn committed:
WIP: proving main goal

pennyannn committed:
Simplify specification

pennyannn committed:
WIP: proving main goal

pennyannn committed:
Further simplify the spec

pennyannn committed:
WIP: proving main goal

pennyannn committed:
Update the spec

pennyannn committed:
WIP: prove main goal

pennyannn committed:
Revert "Update the spec"

This reverts commit 51f0d081d32f75b2927674d4d20f549fc7443027.

nebeid committed:
Moved udiv and second tweak operations to later, WIP: full proof; code before the loop proven

See corresponding https://github.com/aws/aws-lc/pull/2726
Diff of assembly files in both repos:
diff third_party/s2n-bignum/s2n-bignum-to-be-imported/arm/aes/aes-xts-enc.S ../s2n-bignum/arm/aes-xts/aes-xts-armv8.S                                              [25/10/2 | 5:02PM]
4d3
< #include <openssl/asm_base.h>
62c61
<     AARCH64_VALID_CALL_TARGET
---
>     # AARCH64_VALID_CALL_TARGET
67a67
>
136a137

nebeid committed:
Update byte_list_at.

nebeid committed:
Update AES-XTS encrypt spec removing the byte list returned in case of error, matching the decrypt spec updates.

pennyannn committed:
Finished: prove main goal

nebeid committed:
Subgoal 2 of the loop invariant proven.

pennyannn committed:
WIP: prove tail

pennyannn committed:
Restructure proof to add a cut point before cipher stealing and finish up the invariant proof

nebeid committed:
Changed back bytes_to_int128 to its original definition.

nebeid committed:
WIP: subgoal 3 of the loop invariant.

pennyannn committed:
Update the program with respect to AWS-LC PR2734

pennyannn committed:
Finished: verify cipher stealing

nebeid committed:
WIP: subgoal 3 of the loop invariant, arrived at the inductive step.

nebeid committed:
WIP: subgoal 3 of the loop invariant, going through the inductive step.

nebeid committed:
WIP: subgoal 3 of the loop invariant, at the inductive step, reduced the goal by proving the last round in the loop (following the decrypt proof).

nebeid committed:
WIP: subgoal 3 of the loop invariant, at the inductive step, reduced the goal by proving the last 3 rounds in the loop (following the decrypt proof).

nebeid committed:
subgoal 3 of the loop invariant; the inductive step - DONE.

nebeid committed:
Subgoal 4 of main loop invariant: backedge is not taken until i = num_5blocks.

pennyannn committed:
Tried to fix the non-overlapping issue

pennyannn committed:
Fixing some issues

pennyannn committed:
Prepare for bounded proofs

pennyannn committed:
Finished: bounded proofs

pennyannn committed:
Add the final subroutine theorem

nebeid committed:
Assembly changes in cipher-stealing from AWS-LC PR2734 to avoid looping. Subgoal 5: WIP.

nebeid committed:
Subgoal 5: case len mod 0x50 = 0x40.

nebeid committed:
Subgoal 5: case len mod 0x50 = 0x30.

nebeid committed:
Subgoal 5: case len mod 0x50 = 0x20.

nebeid committed:
Subgoal 5: case len mod 0x50 = 0x10.

pennyannn committed:
Thanks for John Harrison, cleaning up the last CHEAT_TAC

nebeid committed:
Subgoal 5: case len mod 0x50 = 0 (wip).

nebeid committed:
Subgoal 5: case len mod 0x50 = 0.

nebeid committed:
Cipher-stealing: case tail = 0.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant up to subgoal 2.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant subgoal 3: WIP.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant back to subgoal 2: WIP.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant subgoal 2: done.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant subgoal 3: symbolic simulation.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant subgoal 3: symbolic simulation completed. Went back to subgoal 2.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant subgoal 3: symbolic simulation completed. Went back to subgoal 2, fix typos and an extra pre-condition.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Reworking subgoal 2 based on l1_curr_len, using the same cipher_stealing_inv as decrypt.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Reworking subgoal 2 (wip), needing a new lemma.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Subgoal 2: Done, pending a new lemma.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Subgoal 3: case i=3 working, now case i = 0 is wip. Subgoal 2 new lemma from Yan working

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Subgoal 3: complete.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Subgoal 4: complete.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Subgoal 5, wip.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Subgoal 5, wip (incomplete subgoals l. 3977 and l. 4485.

nebeid committed:
Cipher-stealing: case tail != 0, loop invariant: Subgoal 5, proved subgoals of previous commit.

nebeid committed:
Cipher-stealing: complete.

nebeid committed:
Less than 2 blocks: with the late W19 setting bug, assumptions are erased in symbolic simulation and cannot match them in TWEAK_TAC.

nebeid committed:
Less than 2 blocks: with the fix setting W19 before branching, no assumptions are erased in symbolic simulation and TWEAK_TAC succeeds.

nebeid committed:
Less than 2 blocks: proof done.

nebeid committed:
Less than 3 blocks: proof done.

nebeid committed:
Less than 4 blocks, less than 5 blocks: proofs done.

nebeid committed:
AES256-XTS encrypt proof: complete.

nebeid committed:
AES256-XTS encrypt proof including stack management: complete.

nebeid committed:
Adjust 1-block proof to the code after the W19 fix.

nebeid committed:
Use the updated tactic name.

pennyannn committed:
Clean up Makefiles and rename files to keep consistency

pennyannn committed:
Refactor the structure of files

pennyannn committed:
Put AES encrypt and decrypt proofs in tutorial

pennyannn committed:
Integrate assembly

nebeid commented:
Thank you, Yan, for combining our 2 long branches together in this PR.

pennyannn committed:
Fix issue in collect-signatures

pennyannn committed:
Add benchmark for AES-XTS

pennyannn committed:
Fix missing sematest for ldrb shifted register variant

pennyannn committed:
Rename AES_KEY as s2n_bignum_AES_KEY

pennyannn committed:
Fix benchmark platform check

pennyannn committed:
Add sematest for strb shifted register no shift

pennyannn committed:
Update signature

pennyannn committed:
Rename aes_hw_xts_encrypt to aes_xts_encrypt and similarly for decrypt

aqjune-aws commented on arm/proofs/utils/aes_decrypt_spec.ml:
```diff
@@ -0,0 +1,143 @@
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0 OR ISC OR MIT-0
+ *)
+
```
Could you add a banner comment to here which will explain this file?

aqjune-aws commented on tools/collect-signatures.py:
```diff
@@ -48,6 +48,10 @@ def __eq__(self, io2):
+  def print(self):
+    print(f"- meminputs: {self.meminputs}")
+    print(f"- memoutputs: {self.memoutputs}")
+    print(f"- temporaries: {self.temporaries}")
```
I guess this function can be removed

aqjune-aws commented on arm/proofs/aes_xts_decrypt.ml:
```diff
@@ -0,0 +1,6190 @@
+    COND_CASES_TAC THENL
+    [ SIMP_TAC[LENGTH_EQ_NIL];
+      SIMP_TAC[] THEN
+      REWRITE_TAC[LET_DEF; LET_END_DEF] THEN
```
FYI - `CONV_TAC (DEPTH_CONV let_CONV)` will work too

nebeid committed:
Merge branch 'main' into aes-xts

nebeid committed:
Added explanation to encrypt tutorial.

nebeid committed:
Put AES encrypt and decrypt proofs in one file in the tutorial, with explanation.

nebeid committed:
Merge remote-tracking branch 'upstream/main' into aes-xts

nebeid committed:
Revert "Put AES encrypt and decrypt proofs in one file in the tutorial, with explanation."

This reverts commit d3035e88f01f4233624a248555a448cc99fa0ec5.

nebeid committed:
Added explanation to the decrypt tutorial.

nebeid committed:
Add banners and remove print function in collect-signatures.py.

nebeid committed:
Put AES encrypt and decrypt proofs in one file in the tutorial (2nd attempt).

nebeid committed:
Merge remote-tracking branch 'upstream/main' into aes-xts

nebeid committed:
Put specs for list and tweak calculation in a common spec file.

Add ASCII drawing to explain the cipher-stealing loop invariant.
Address comments on decrypt proof.

nebeid committed:
Update the name of the encrypt subroutine theorem to match the decrypt one.

nebeid committed:
Merge branch 'main' into aes-xts

nebeid committed:
Accommodate files that can work on Windows.

nebeid committed:
Merge branch 'main' into aes-xts

nebeid committed:
Replace non-ASCII characters with ASCII equivalents

jargh commented:
The big picture looks great - the code is right, the spec is nicely structured and the proofs all work. Just a few routine "s2n-bignum house style" things:
1. It needs rebasing by now - that should be routine I hope
2. Since this PR was first submitted, the canonical boilerplate standard for s2n-bignum functions has changed slightly, in particular acquiring CFI directives, preferably via the wrapping macros
3. Having just known-answer tests is a bit of a departure from the usual s2n-bignum testing infrastructure where we prefer to have a general reference to test against (though KATs are nice too).

With a bit of help from Claude, here are a couple of proposed patch files to address 2 and 3 - see what you think :-)
[pr331_cfi_fix.patch](https://github.com/user-attachments/files/27079104/pr331_cfi_fix.patch)
[pr331_ref_tests.patch](https://github.com/user-attachments/files/27079105/pr331_ref_tests.patch)

jargh committed:
Merge remote-tracking branch 'awslabs/main' into aes-xts

jargh committed:
Add CFI directives and macOS-compatible labels to AES-XTS assembly

- Switch to _internal_s2n_bignum_arm.h for CFI/metadata macros
- Add CFI_START, CFI_RET, CFI_DEC_SP/CFI_INC_SP, CFI_STACKSAVE2X/
  CFI_STACKLOAD2 for correct stack unwinding
- Add S2N_BN_FUNCTION_TYPE_DIRECTIVE and S2N_BN_SIZE_DIRECTIVE
- Rename .Lxxx local labels to Lxxx (L-prefix convention for Mach-O)

jargh committed:
Add reference C implementation and random test harness for AES-XTS

- Add tests/ref_aes_xts.c: standalone AES-256-XTS implementation
  following FIPS 197 and IEEE 1619, with encrypt/decrypt and
  ciphertext stealing for non-block-aligned inputs
- Add random-input tests exercising both encrypt and decrypt across
  a range of input lengths (16 to 300+ bytes), verifying round-trip
  correctness and consistency with the optimized assembly

This was entirely written by Claude Opus 4.6.

jargh committed:
Fix missing closing brace in ref_aes_xts.c

jargh reviewed (approved):
All looks good to me now, thanks :-)
